Fixes #8945: Cleanup log symlinks on remove & dead log symlinks

jimmidyson opened the pull request:
@satnam6502 Care to review?

jimmidyson committed:
Fixes #8945: Cleanup log symlinks on remove & dead log symlinks

satnam6502 commented on pkg/kubelet/container_gc.go:
```diff
@@ -157,6 +170,19 @@ func (cgc *realContainerGC) GarbageCollect() error {
+	// from a k8s version without proper log symlink cleanup
+	logSymlinks, _ := filepath.Glob(path.Join(cgc.containerLogsDir, "*.log"))
+	for _, logSymlink := range logSymlinks {
+		_, err = os.Stat(logSymlink)
```
Minor, but you could inline this into the `if` statement on the next line.

satnam6502 commented:
Thanks!
CC @a-robinson 
TLs: @dchen1107 @thockin @roberthbailey @quinton-hoole 
I shall defer to the triage committee about whether this is post 1.0.

satnam6502 commented:
@k8s-bot ok to test

jimmidyson commented:
@satnam6502 Updated as requested.

I do think this is important for 1.0 - otherwise loads of dead symlinks as containers are cleaned up & log symlinks not. Wasted inodes & means fluentd logger has to watch dead symlinks - not sure of the impact of that, but not very pretty.

satnam6502 commented:
I agree with @jimmidyson -- it would be great to get this in if at all possible.
Risk: medium-low-ish.

dchen1107 commented:
I agreed to include this for 1.0.

dchen1107 commented:
LGTM. 

 cc/ @lavalamp @bgrant0607 @thockin or @brendandburns 

I inclined to include this to 1.0 release to avoid , but ok to post it for 1.X release too.

quinton-hoole commented on pkg/kubelet/container_gc.go:
```diff
@@ -79,6 +86,12 @@ type containerGCInfo struct {
 	createTime time.Time
+
+	// Pod name
+	podName string
```
What about the namespace?

jimmidyson replied:
From https://github.com/GoogleCloudPlatform/kubernetes/blob/master/pkg/kubelet/dockertools/docker.go#L273 this podName is in 'namespace_podName' format.

quinton-hoole replied:
Please add a comment, as that's not at all obvious.

lavalamp replied:
Also change the name of the variable. Don't call it a name if it includes a namespace. :)

quinton-hoole commented on pkg/kubelet/container_gc.go:
```diff
@@ -157,6 +170,18 @@ func (cgc *realContainerGC) GarbageCollect() error {
 
+	// Remove dead symlinks - should only happen on upgrade
+	// from a k8s version without proper log symlink cleanup
+	logSymlinks, _ := filepath.Glob(path.Join(cgc.containerLogsDir, "*.log"))
```
"*.log" is fragile here, as there's nothing explicit keeping it in sync with the log file names/symlinks themselves.  Best make the file extension a shared constant to keep the log file names and this wildcard in sync, or else elevate this literal string to a named constant at the top of the file, with reciprocal comments to keep it in sync with the log file names.

quinton-hoole commented on pkg/kubelet/container_gc.go:
```diff
@@ -157,6 +170,18 @@ func (cgc *realContainerGC) GarbageCollect() error {
+	// from a k8s version without proper log symlink cleanup
+	logSymlinks, _ := filepath.Glob(path.Join(cgc.containerLogsDir, "*.log"))
+	for _, logSymlink := range logSymlinks {
+		if _, err = os.Stat(logSymlink); os.IsNotExist(err) {
```
As far as I can determine, this won't work, as the stat() call on the dead symlink will succeed?

e.g. (and yes, the results are the same under linux):

```
4 lrwxr-xr-x 1 quinton eng   8 Jul  7 13:43 test.sym -> test.txt
0 -rw-r--r-- 1 quinton eng   0 Jul  7 13:42 test.txt
quinton-macbookpro:symlink-test quinton$ stat test.sym
  File: ‘test.sym’ -> ‘test.txt’
  Size: 8         	Blocks: 8          IO Block: 4096   symbolic link
Device: 1000004h/16777220d	Inode: 64971777    Links: 1
Access: (0755/lrwxr-xr-x)  Uid: (157215/ quinton)   Gid: ( 5000/     eng)
Access: 2015-07-07 13:43:05.000000000 -0700
Modify: 2015-07-07 13:43:05.000000000 -0700
Change: 2015-07-07 13:43:05.000000000 -0700
 Birth: 2015-07-07 13:43:05.000000000 -0700
quinton-macbookpro:symlink-test quinton$ rm test.txt
quinton-macbookpro:symlink-test quinton$ stat test.sym
  File: ‘test.sym’ -> ‘test.txt’
  Size: 8         	Blocks: 8          IO Block: 4096   symbolic link
Device: 1000004h/16777220d	Inode: 64971777    Links: 1
Access: (0755/lrwxr-xr-x)  Uid: (157215/ quinton)   Gid: ( 5000/     eng)
Access: 2015-07-07 13:43:05.000000000 -0700
Modify: 2015-07-07 13:43:05.000000000 -0700
Change: 2015-07-07 13:43:05.000000000 -0700
 Birth: 2015-07-07 13:43:05.000000000 -0700
```

Am I missing something?

jimmidyson replied:
What's your test code like? are you sure you're using os.Stat & not os.Lstat? former should follow symlink. worked in my tests earlier but will retry tomorrow to make sure.

quinton-hoole replied:
I didn't write test code, I just used the standard 'stat' command line tool, as per my comment above (without the -L option to follow links).

jimmidyson replied:
ah sorry I didn't know about stat cmd line tool. anyway, os.Stat follows symlinks so this is why it works. os.Lstat would be the equivalent of what you did with stat.

quinton-hoole replied:
Yes, sounds like you're all good. Just checking.

quinton-hoole commented on pkg/kubelet/container_gc.go:
```diff
@@ -179,6 +204,11 @@ func (cgc *realContainerGC) removeOldestN(containers []containerGCInfo, toRemove
 		if err != nil {
 			glog.Warningf("Failed to remove dead container %q: %v", containers[i].name, err)
 		}
+		symlinkPath := path.Join(cgc.containerLogsDir, fmt.Sprintf("%s_%s-%s.log", containers[i].podName, containers[i].containerName, containers[i].id))
```
Similar comment to above.  What keeps this in sync with the actual log file names?  It seems that you need a common utility function to map from container->logfilename.

jimmidyson replied:
Good idea

jimmidyson commented:
@quinton-hoole @lavalamp Addressed your comments. I can't access the shippable build for some reason to see why it's failing... do you need to allow me to access it? Would be good if the builds were open for all to see.

jimmidyson commented:
@satnam6502 Not sure why the builds failed - seems like Kibana didn't start which is totally unrelated to this PR. Can we re-run tests?

satnam6502 commented:
@k8s-bot ok to test

satnam6502 commented:
The GCE failure wad due to the build timing out.
@k8s-bot ok to test

quinton-hoole commented:
@k8s-bot test this again please

quinton-hoole commented:
Code LGTM.  I've asked both Shippable (unit and integration tests) and Jenkins (end-to-end tests) to rerun, as both failures are flakes unrelated to this PR.

quinton-hoole commented:
I'll add ok-to-merge as soon as the tests pass.

jimmidyson commented:
@quinton-hoole Thanks!
